src/future/batch_semaphore.rs

```diff
@@ -179,7 +358,18 @@ impl BatchSemaphore {
     /// If there aren't enough permits, returns `Err(TryAcquireError::NoPermits)`
     pub fn try_acquire(&self, num_permits: usize) -> Result<(), TryAcquireError> {
         let mut state = self.state.borrow_mut();
-        state.acquire_permits(num_permits)
+        state.acquire_permits(num_permits).map_err(|err| {
+            // Conservatively, the requester causally depends on the
+            // last successful acquire.
+            // TODO: This is not precise, but `try_acquire` causal
```

Comment:
Think you should add the commentary on what are the options, and why this choice is not as precise as the alternatives (of tracking clocks for all current permit holders).